File: src/sensor_nivel.py
class SensorNivel:
    def __init__(self, tag: str, valor: float, unidade: str = ""):
        self._tag = tag
        self._valor = valor
        self._unidade = unidade
        self._ativo = False
        self._total_leituras = 0

    @property
    def tag(self):
        return self._tag

    @property
    def valor(self):
        return self._valor

    @property
    def unidade(self):
        return self._unidade

    @property
    def ativo(self):
        return self._ativo

    @property
    def total_leituras(self):
        return self._total_leituras

    def ativar(self):
        self._ativo = True

    def desativar(self):
        self._ativo = False

    def registrar_leitura(self, novo_valor):
        if not self._ativo:
            return False

        if novo_valor < 0 or novo_valor > 100:
            return False

        self._valor = novo_valor
        self._total_leituras += 1

        return True

    def resumo(self) -> str:
        sufixo = f" {self._unidade}" if self._unidade else ""
        estado = "ativo" if self._ativo else "inativo"

        return f"{self._tag}: {self._valor:g}{sufixo} | {estado} | leituras: {self._total_leituras}"

File: src/sensor_nivel.cpp
#include "sensor_nivel.hpp"

#include <iomanip>
#include <sstream>
#include <utility>

namespace {
std::string formatarValor(double valor) {
    std::ostringstream saida;
    saida << std::setprecision(12) << valor;
    return saida.str();
}
}

SensorNivel::SensorNivel(std::string tagInicial, double valorInicial, std::string unidadeInicial)
    : _tag(std::move(tagInicial)),
      _valor(valorInicial),
      _unidade(std::move(unidadeInicial)),
      _ativo(false),
      _totalLeituras(0) {
}

std::string SensorNivel::tag() const {
    return _tag;
}

double SensorNivel::valor() const {
    return _valor;
}

std::string SensorNivel::unidade() const {
    return _unidade;
}

bool SensorNivel::estaAtivo() const {
    return _ativo;
}

void SensorNivel::ativar() {
    _ativo = true;
}

void SensorNivel::desativar() {
    _ativo = false;
}

bool SensorNivel::registrarLeitura(double novoValor) {
    if (!_ativo) {
        return false;
    }

    if (novoValor < 0 || novoValor > 100) {
        return false;
    }

    _valor = novoValor;
    _totalLeituras++;

    return true;
}

int SensorNivel::totalLeituras() const {
    return _totalLeituras;
}

std::string SensorNivel::resumo() const {
    const std::string sufixo = _unidade.empty() ? "" : " " + _unidade;
    const std::string estado = _ativo ? "ativo" : "inativo";

    return _tag + ": " + formatarValor(_valor) + sufixo +
           " | " + estado +
           " | leituras: " + std::to_string(_totalLeituras);
}